pkg/sys: Create user namespace as the container's initial user namesp…

halaney opened the pull request:
…ace user

Right now containerd uses os.StartProcess() to create a dummy ptraced process via a fork/exec in a new user namespace, with the uid_map/gid_map setup.

This doesn't work so well with user limits[0] in the kernel, at least if you expect the container user to have separate isolation from the host's "normal users". The kernel's ucount mechanism forms a ucount tuple of (ns, kuid_t) for each user limit. Say containerd runs as uid 0 in the initial user namespace, and the container runs as uid 100000 from the inital user namespace point of view.

When the container tries to do things like inotify_add_watch(), the kernel:

    1. First verifies that the ucount(container user namespace, 100000) doesn't exceed its "per user per user namespace limits" and increments the counter there. This limit is set by /proc/sys/user/max_inotify_watches when in the namespace
    2. Then walks up to the ucount who created this namespace, ucount(initial user namespace, 0) and increments their counter as well ensuring they don't exceed their limit. In our example this is the initial user namespace, this limit is set by /proc/sys/fs/inotify/max_user_watches as well as /proc/sys/user/max_inotify_watches in that case.

This is done so a user can't escape per user limits by creating a user namespace and running as different users in that user namespace. The accounting always rolls back up to the user who created the user namespace to ensure this, checking limits at each layer.

This means if you have a rogue container they can consume all of the containerd user's inotify limits. In practice this means global root is in danger of being denied usage of resources due to the container, while other less important users are still well within their budget! In the inotify case for example systemd will fail to start many new services due to this.

Let's instead create the user namespace *as the container's initial user namespace user*. This means that all the attribution for these user limits rolls up to this unimportant user, preventing one container from exhausting global root's resources, and further isolating each container from each other (since each pod in k8s runs as its own disjoint set of users in the initial user namespace).

This is a bit annoying to do in golang. The best option seems to be what we have here:

    1. Lock OS thread
    2. Manually setresuid() (seteuid implementation in golang mimics glibc's and sets the euid for all of the threads, which causes issues for other threads running at the same time)
    3. Create user namespace
    4. Undo the setresuid()
    5. Update the uid_map/gid_map as root now (can't do this anymore as part of (3) since to update the map you must have CAP_SETUID which the container user will not have)

With this in place you can verify the user namespace is owned by the proper uid via ioctl(fd, NS_GET_OWNER_UID).

[0]: https://docs.kernel.org/admin-guide/sysctl/user.html#documentation-for-proc-sys-user

halaney commented:
Hmmm CI failures are related for sure
```
=== FAIL: pkg/sys TestUnshareAfterEnterUserns/user_namespace_ownership (0.00s)
    unshare_linux_test.go:168: 
        	Error Trace:	/home/runner/work/containerd/containerd/pkg/sys/unshare_linux_test.go:168
        	Error:      	Received unexpected error:
        	            	failed to start noop process for unshare: fork/exec /proc/self/exe: operation not permitted
        	Test:       	TestUnshareAfterEnterUserns/user_namespace_ownership
```
and then in the dmesg dump
```
2025-09-22T19:52:19.6362000Z [Mon Sep 22 19:51:44 2025] audit: type=1400 audit(1758570704.213:137): apparmor="AUDIT" operation="userns_create" class="namespace" info="Userns create - transitioning profile" profile="unconfined" pid=47611 comm="podsandbox.test" requested="userns_create" target="unprivileged_userns"
2025-09-22T19:52:19.6363512Z [Mon Sep 22 19:51:44 2025] audit: type=1400 audit(1758570704.214:138): apparmor="DENIED" operation="capable" class="cap" profile="unprivileged_userns" pid=47619 comm="podsandbox.test" capability=21  capname="sys_admin"
```

Maybe something to do with unprivileged user namespace [stuff](https://discourse.ubuntu.com/t/spec-unprivileged-user-namespace-restrictions-via-apparmor-in-ubuntu-23-10/37626?_gl=1*1m8rgvp*_gcl_au*MTEwNTI4ODU5My4xNzU4NTc3MTMw)? Need to look more at that, not very apparmor savvy

rata commented on pkg/sys/unshare_linux.go:
```diff
@@ -87,8 +67,21 @@ func UnshareAfterEnterUserns(uidMap, gidMap string, unshareFlags uintptr, f func
 		pidfdWaitid(pidfd)
 	}()
 
+	// Now we can write the uid/gid mappings
+	uidMapPath := fmt.Sprintf("/proc/%d/uid_map", pid)
+	uidMapContent := fmt.Sprintf("%d %d %d\n", uidMaps[0].ContainerID, uidMaps[0].HostID, uidMaps[0].Size)
+	if err := os.WriteFile(uidMapPath, []byte(uidMapContent), 0644); err != nil {
+		return fmt.Errorf("failed to write UID mapping: %w", err)
+	}
+
+	gidMapPath := fmt.Sprintf("/proc/%d/gid_map", pid)
+	gidMapContent := fmt.Sprintf("%d %d %d\n", gidMaps[0].ContainerID, gidMaps[0].HostID, gidMaps[0].Size)
+	if err := os.WriteFile(gidMapPath, []byte(gidMapContent), 0644); err != nil {
+		return fmt.Errorf("failed to write GID mapping: %w", err)
+	}
```
So the process creation is not failing in any way, go just fails to set the mapping, ignores it and doesn't return any errors? That seems surprising on itself too

rata replied:
Oh, but this unprivileged user of course can't map IDs other than itself. So we need this to be done from the root process

rata commented:
@halaney right, I forgot that ubuntu blocked unprivileged userns creation by default. Not sure if any other popular distro is doing that still or not. But it's definitely something to take into account. And _I_ wouldn't go the route of an apparmor profile, it doesn't seem nice and it is very ubuntu specific. Some other distros have used other means to block it (debian had a sysctl in the past).

I guess this needs more thoughts :-(

rata commented on pkg/sys/unshare_linux.go:
```diff
@@ -137,6 +176,54 @@ func parseIDMapping(mapping string) ([]syscall.SysProcIDMap, error) {
+		return -1, -1, fmt.Errorf("failed to set effective UID: %w", errno)
+	}
+	defer func() {
+		syscall.RawSyscall(unix.SYS_SETRESUID, ^uintptr(0), uintptr(originalEUID), ^uintptr(0))
```
What if this fail? I guess containerd will be in a very weird state, as one OS thread will be running unprivileged and things will fail unexpectedly.

Is there a better alternative than all of the test in a goroutine with OSLockThread and, then, if it fails we just finish the goroutine without finishing the lock? In that case the go runtime will kill the OS thread:

See https://pkg.go.dev/runtime#LockOSThread:

> LockOSThread wires the calling goroutine to its current operating system thread. The calling goroutine will always execute in that thread, and no other goroutine will execute in it, until the calling goroutine has made as many calls to [UnlockOSThread](https://pkg.go.dev/runtime#UnlockOSThread) as to LockOSThread. If the calling goroutine exits without unlocking the thread, the thread will be terminated.

halaney commented on pkg/sys/unshare_linux.go:
```diff
@@ -20,13 +20,56 @@ import (
+	nobodyUID := 65534
+	var pidfd int
+
+	_, pidfd, err := startProcessWithUserNamespace(nobodyUID, syscall.CLONE_NEWIPC)
```
I picked a random one namespace here, but in general I think we need to properly test the unshare path as that's what apparmor starts to choke on after entering an unprivileged user namespace.

halaney commented on pkg/sys/unshare_linux.go:
```diff
@@ -137,6 +176,54 @@ func parseIDMapping(mapping string) ([]syscall.SysProcIDMap, error) {
+	runtime.LockOSThread()
+	defer runtime.UnlockOSThread()
+
+	if _, _, errno := syscall.RawSyscall(unix.SYS_SETRESUID, ^uintptr(0), uintptr(targetHostUID), ^uintptr(0)); errno != 0 {
```
If I understand correctly its safe to do this in that case as well, although unnecessary. 

To me I'd rather avoid the conditional and do this the same no matter what instead of introducing an edge case but open to differing opinions there.

halaney commented:
> @halaney friendly ping?

Thanks @rata, finally found some time to return to this and other PRs :D 

I just pushed a new version addressing your comments, I'll let you resolve them as you re-review. The goroutine defer stuff is a little funky but I think its the best approach

halaney commented on pkg/sys/unshare_linux.go:
```diff
@@ -137,6 +176,95 @@ func parseIDMapping(mapping string) ([]syscall.SysProcIDMap, error) {
+		defer func() {
+			_, _, restoreErrno = syscall.RawSyscall(unix.SYS_SETRESUID, ^uintptr(0), uintptr(originalEUID), ^uintptr(0))
+			if restoreErrno != 0 {
+				if proc != nil {
```
> A function that inside has an anonymous function, that inside has a defer function, that inside does two nested ifs. Let's simplify this.

:P when you put it like that... yes, let me revisit. I was trying to get fancy with clean ups using defers, but I think it might just be a lot more clear to do all the clean up in the error handling directly in this case.

rata commented:
I'd suggest something like this: https://github.com/containerd/containerd/pull/12317#pullrequestreview-3589772999

halaney commented on pkg/sys/unshare_linux.go:
```diff
@@ -20,13 +20,56 @@ import (
+
+	unix.PidfdSendSignal(pidfd, unix.SIGKILL, nil, 0)
+	pidfdWaitid(pidfd)
+	unix.Close(pidfd)
```
You're right, `proc.Kill()` *does* seem to use pidfd for signaling when it has that!
https://github.com/golang/go/blob/fd45d70799853ef5980a559e206353574108d8da/src/os/exec_posix.go#L76
https://github.com/golang/go/blob/1843cfbcd6bd0e1f7650c663c6c4608bc1cd8829/src/os/exec_unix.go#L86

I can just change to `proc.Kill()/proc.Wait()` after we resolve other review convos

rata replied:
Cool! Yeah, the rest is fine, thanks for all the answers! Let's simplify this and it's ready from my pov :)

halaney replied:
@rata, I'm a bit of a fool and didn't think the logistics of that out. We don't have a reference to the `proc` here to do `proc.Kill()`. If I make it so `startProcessWithUserNamespace()` returns a proc... then we're starting to get into the territory of "rework this all to use `proc` directly and rely on the go runtime to use pidfd's instead of /proc/ interfaces" as was discussed in the other [thread](https://github.com/containerd/containerd/pull/12317/files#r2653366372).

The more I think about it I like doing that, but I do still think that's an independent topic that we should refactor and review separately. So for now I'm keeping this as is to match usage in `UnshareAfterEnterUserns`.

Lemme know if you disagree and I'll expand the scope of the PR to include that work too, but I think this is already a big enough topic to review

rata replied:
Nah, looks fine because this only work when pidfd works. Thanks!

rata commented on pkg/sys/unshare_linux.go:
```diff
@@ -20,13 +20,56 @@ import (
+	// that we can indeed make a user namespace, then unshare into another namespace
+	// as the target user without hitting any sort of apparmor restrictions
+	_, pidfd, err := startProcessWithUserNamespace(nobodyUID, syscall.CLONE_NEWIPC)
+	if err != nil {
```
Oh, okay, no more tweaks needed here, then! :)

rata commented on pkg/sys/unshare_linux.go:
```diff
@@ -137,6 +177,82 @@ func parseIDMapping(mapping string) ([]syscall.SysProcIDMap, error) {
+	if pidfd == -1 {
+		proc.Kill()
+		proc.Wait()
+		return -1, -1, fmt.Errorf("kernel doesn't support CLONE_PIDFD")
```
Okay, so let's keep it as is! :)

rata reviewed (approved):
LGTM, thanks! Left a simple small comment.

@fuweid can you take a look? You worked on the pidfd stuff, it will be great ifyou can take a look.

rata commented on pkg/sys/unshare_linux.go:
```diff
@@ -137,6 +177,82 @@ func parseIDMapping(mapping string) ([]syscall.SysProcIDMap, error) {
+// is *not* setresuid()'ed back to the original user. In this case the thread should be killed off
+// by the caller to avoid using a thread in a bad state
+func startProcessWithUsernsLocked(targetHostUID int, unshareFlags uintptr) (int, int, error) {
+	originalEUID := os.Geteuid()
```
Agree, let's use -1 that is documented :)

fuweid commented on pkg/sys/unshare_linux.go:
```diff
@@ -137,6 +177,88 @@ func parseIDMapping(mapping string) ([]syscall.SysProcIDMap, error) {
+		originalEUID = os.Geteuid()
+	}
+
+	if _, _, errno := syscall.RawSyscall(unix.SYS_SETRESUID, ^uintptr(0), uintptr(targetHostUID), ^uintptr(0)); errno != 0 {
```
Based on doc https://www.man7.org/linux/man-pages/man7/capabilities.7.html#:~:text=Effect%20of%20user,ambient%20capability%0A%20%20%20%20%20%20%20%20%20%20sets., after setresuid, thread's capability will be cleared.

> If one or more of the real, effective, or saved set user IDs
          was previously 0, and as a result of the UID changes all of
          these IDs have a nonzero value, then all capabilities are
          cleared from the permitted, effective, and ambient capability
          sets.

Based on my local test, this change doesn't work. If I understand it correctly, we need to restore capability after retsetuid. For example,


```go
@@ -216,16 +216,28 @@ func startProcessWithUserNamespace(targetHostUID int, unshareFlags uintptr) (int
 func startProcessWithUsernsLocked(targetHostUID int, unshareFlags uintptr) (int, int, error) {
        // -1 means no change in setresuid()
        originalEUID := -1
+       var originalCaps capSnapshot

        if targetHostUID != -1 {
                // We're changing users, so we need to figure out what user to return to
                originalEUID = os.Geteuid()
+               var err error
+               originalCaps, err = getCaps()
+               if err != nil {
+                       return -1, -1, fmt.Errorf("failed to read capabilities: %w", err)
+               }
        }

        if _, _, errno := syscall.RawSyscall(unix.SYS_SETRESUID, ^uintptr(0), uintptr(targetHostUID), ^uintptr(0)); errno != 0 {
                return -1, -1, fmt.Errorf("failed to set effective UID: %w", errno)
        }

+       if targetHostUID != -1 {
+               if err := setCaps(originalCaps); err != nil {
+                       return -1, -1, fmt.Errorf("failed to restore capabilities: %w", err)
+               }
+       }
+
        var pidfd int
```

rata replied:
> No sure we should rely security profile to do that.

@fuweid what do you mean?

rata replied:
Awesome! And if I understand correctly, we don't need to set capabilities again when switching back to UID 0 as (from capabilities manpage):
>        •  If the effective user ID is changed from nonzero to 0, then the permitted set is copied to the effective set.

I wonder how the effective set or the ambient set are affected by this.

@halaney can you confirm that all the capabilities set (effective, ambient, etc.) are the same before we switch from UID 0 to after we switch back to 0?

It's getting a little hairy and capabilities are more complex that I'd like. Another option is to always let the OS thread die. If we don't get this right, the debug of this will be SO painful

halaney replied:
Yeah agreed on the capset comment about not changing ambient/bounded (honestly forgot those existed, we're so in the weeds here with the syscalls I love it).

I scraped `/proc/self/status` for those bounded/ambient caps from the locked thread since it seems hard to get otherwise via `prctrl()` -- they're not changed at all 
```
2026/01/14 10:08:59 INFO HALANEY: ORIGINAL CAPS
2026/01/14 10:08:59 INFO Capabilities: effective=0x000001ffffffffff permitted=0x000001ffffffffff inheritable=0x0000000800000000 bounding=0x000001ffffffffff ambient=0x0000000000000000
2026/01/14 10:08:59 INFO HALANEY: CAPS AFTER SETRESUID (before setcaps)
2026/01/14 10:08:59 INFO Capabilities: effective=0x0000000000000000 permitted=0x000001ffffffffff inheritable=0x0000000800000000 bounding=0x000001ffffffffff ambient=0x0000000000000000
2026/01/14 10:08:59 INFO HALANEY: CAPS AFTER SETRESUID + SETCAPS
2026/01/14 10:08:59 INFO Capabilities: effective=0x000001ffffffffff permitted=0x000001ffffffffff inheritable=0x0000000800000000 bounding=0x000001ffffffffff ambient=0x0000000000000000
2026/01/14 10:08:59 INFO HALANEY: FINAL CAPS
2026/01/14 10:08:59 INFO Capabilities: effective=0x000001ffffffffff permitted=0x000001ffffffffff inheritable=0x0000000800000000 bounding=0x000001ffffffffff ambient=0x0000000000000000
```

rata commented:
Friendly ping? :)
